Neos.Neos/Classes/FrontendRouting/EventSourcedFrontendNodeRoutePartHandler.php

```
@@ -228,22 +238,25 @@ public function matchWithParameters(&$requestPath, RouteParameters $parameters)
     /**
      * @param string $uriPath
      * @param DimensionSpacePoint $dimensionSpacePoint
-     * @return MatchResult
+     * @return MatchResult|false
      * @throws NodeNotFoundException
      */
     private function matchUriPath(
         string $uriPath,
         DimensionSpacePoint $dimensionSpacePoint,
         SiteDetectionResult $siteDetectionResult,
         ContentRepository $contentRepository
-    ): MatchResult {
+    ): MatchResult|false {
```

Do we really need the union type here? We could also throw an exception?  Or make the return type nullable?

---

We had some discussion about what to return. The exception seemed wrong, false is more explicit.
Is there a  guideline what we prefer to have for the core?
Some also prefer to have types for the negative result.

---

It's hard to come up with absolute heuristics, but as a rule of thumb: we should not use exceptions for control flow.
So in this case:
If you try to create a link for a node that is removed, it is fine to throw an exception because that case is not to be expected.

if the node exists but it does not match the `nodeType` filter, that is not an exceptional case because we are reusing the same class for all node-related routes so it is expected that a couple of them don't match the filter.

The line is blurry but essentially throwing an exception has an overhead because the stack trace has to be generated. So personally I prefer the union type in this case even though it is a sign of bad abstraction (but it's just a private method anyways)